### PR TITLE
Added support for authentication codes for the VM service.

### PR DESCRIPTION
**Pull request by bkonyi**
Previously, a valid web socket connection would use the following URI:

ws://127.0.0.1/ws

Now, by default, the VM service requires a connection to be made with a
URI similar to the following:

ws://127.0.0.1:8181/Ug_U0QVsqFs=/ws

where Ug_U0QVsqFs is an authentication code generated and shared by the
service.

This behavior can be disabled with the --disable-service-auth-codes
flag.

Depends on flutter/engine#8594

### CODE REVIEW

File: dev/devicelab/lib/framework/utils.dart [hunk 1]
```diff
@@ -541,6 +541,20 @@ int parseServicePort(String line, {
...
+/// Tries to extract a Uri from the string.
+///
+/// The `prefix`, if specified, is a regular expression pattern and must not contain groups.
+///
+/// The `multiLine` flag should be set to true if `line` is actually a buffer of many lines.
+Uri parseServiceUri(String line, {
+  String prefix = 'An Observatory debugger .* is available at: ',
+  bool multiLine = false,
+}) {
+  final RegExp r = RegExp('$prefix((http|\/\/)[a-zA-Z0-9:/=\.\\[\\]]+)');
```
[the remark this answers is not in the document]
Comment (bkonyi) [reply dev/devicelab/lib/framework/utils.dart]: Fair point, I'll make the change.

FYI, this code was mostly copied from `parseServicePort` above with the actual regular expression grabbed from [flutter_tools](https://github.com/flutter/flutter/blob/2d3ff10d62a1f1fdcbdbe92cbcc5a5594f95d500/packages/flutter_tools/lib/src/protocol_discovery.dart#L63). It seems like this anti-pattern of creating RegExps once happens in quite a few places throughout the codebase.

File: dev/devicelab/lib/framework/utils.dart [hunk 2]
```diff
@@ -541,6 +541,20 @@ int parseServicePort(String line, {
   return match == null ? null : int.parse(match.group(2));
 }
 
+/// Tries to extract a Uri from the string.
+///
+/// The `prefix`, if specified, is a regular expression pattern and must not contain groups.
+///
+/// The `multiLine` flag should be set to true if `line` is actually a buffer of many lines.
+Uri parseServiceUri(String line, {
```
[the remark this answers is not in the document]
Comment (bkonyi) [reply dev/devicelab/lib/framework/utils.dart]: Done.

File: dev/devicelab/lib/framework/utils.dart [hunk 3]
```diff
@@ -145,14 +144,14 @@ void section(String title) {
...
-    output += '═';
+  if (output.length == 79) output += '═';
   print('\n\n$output\n');
 }
 
 Future<String> getDartVersion() async {
   // The Dart VM returns the version text to stderr.
-  final ProcessResult result = _processManager.runSync(<String>[dartBin, '--version']);
+  final ProcessResult result =
+      _processManager.runSync(<String>[dartBin, '--version']);
```
Comment (jonahwilliams) [inline dev/devicelab/lib/framework/utils.dart]: Nit: here and below, retain the existing formatting for unrelated code

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 4]
```diff
@@ -199,11 +227,18 @@ class AttachCommand extends FlutterCommand {
       }
     } else {
       usesIpv6 = ipv6;
-      final int localPort = observatoryPort
-        ?? await device.portForwarder.forward(devicePort);
-      observatoryUri = usesIpv6
-        ? Uri.parse('http://[$ipv6Loopback]:$localPort/')
-        : Uri.parse('http://$ipv4Loopback:$localPort/');
+      if (debugUri != null) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/attach.dart]: This is taking on a life of its own, lets refactor it into a method

**Comment by jonahwilliams** [comment]
So the changes are:

- Allow specifying the full uri for attach
- Disable for iOS/Android (?)

I would expect that we need an update to the protocol discovery logic too, and potentially the mdns code

**Comment by bkonyi** [comment]
> So the changes are:
> 
> * Allow specifying the full uri for attach
> * Disable for iOS/Android (?)
> 
> I would expect that we need an update to the protocol discovery logic too, and potentially the mdns code

The protocol discovery logic seems to work as is. Pretty sure it just parses the full URI from some buffer and didn't just go hunting for a port. I'll double check the MDNS code to make sure it does the same.

File: packages/flutter_tools/lib/src/commands/attach.dart (hunk 4 above)
Comment (bkonyi) [reply packages/flutter_tools/lib/src/commands/attach.dart]: Done.

File: dev/devicelab/lib/framework/utils.dart (hunk 3 above)
Comment (bkonyi) [reply dev/devicelab/lib/framework/utils.dart]: Done.

**Comment by jonahwilliams** [comment]
I'm actually not sure how the adb port forwarding logic will handle this, I'm surprised if it works as is.

**Comment by bkonyi** [comment]
> I'm actually not sure how the adb port forwarding logic will handle this, I'm surprised if it works as is.

I've done some testing on an Android device and it seems to behave as expected * shrugs *.

I've been looking at the mdns code but I'm not exactly sure what's going on there. I'm guessing @dnfield has more context here?

**Comment by jonahwilliams** [comment]
I mean, if it works it works, I'm just surprised :)

**Comment by jonahwilliams** [comment]
He'll be back next week.

My concern is more in the cost of requiring to pass the uri at all. For the android case, we should make sure we scrape the token out of the logs in the attach case so that it doesn't need to be specified.

For iOS, we'll need to find a way to discover the token as well (does this defeat the point?) or else we need to go back to log parsing.

**Comment by bkonyi** [comment]
> My concern is more in the cost of requiring to pass the uri at all. For the android case, we should make sure we scrape the token out of the logs in the attach case so that it doesn't need to be specified.

The only time you'll need to specify the URI is when you would have needed to provide `--debugPort` previously. If you run `flutter attach` before running the application, it will pick up the right URI from the logs.

> For iOS, we'll need to find a way to discover the token as well (does this defeat the point?) or else we need to go back to log parsing.

Yeah, I honestly have no clue how MDNS works or what it does. I'm not sure how we'll be able to get the token as it's part of the path, which I doubt is included anywhere. What is MDNS used for anyway?

**Comment by dnfield** [comment]
The mdns logic gets the uri from the service isolate.  Same place as the logs.

It's used for flutter attach on Mac/iOS.

File: dev/devicelab/lib/framework/utils.dart [hunk 5]
```diff
@@ -529,19 +529,43 @@ String extractCloudAuthTokenArg(List<String> rawArgs) {
   return token;
 }
 
+final RegExp obsRegExp =
```
Comment (zanderso) [inline dev/devicelab/lib/framework/utils.dart]: These can be library private e.g. '_obsRegExp'

Comment (bkonyi) [reply dev/devicelab/lib/framework/utils.dart]: Done.

**Comment by bkonyi** [comment]
I've added support for discovering the authentication code via the MDNS text record on iOS (see https://github.com/flutter/engine/pull/8625 for the relevant engine changes), so that should resolve the last major issue. Tested to work on both iOS device and simulator.

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 6]
```diff
@@ -399,7 +443,32 @@ class MDnsObservatoryPortDiscovery {
...
                    '- using first one (${srv.first.port}).');
       }
-      return srv.first.port;
+      printStatus('Checking for authentication code for $domainName');
+      final List<TxtResourceRecord> txt = await client
+        .lookup<TxtResourceRecord>(
+            ResourceRecordQuery.text(domainName),
+        )
+        .toList();
+      String authCode = '';
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/attach.dart]: How about:

```
if (txt.isEmpty) {
  return MDnsObservatoryDiscoveryResult(srv.first.port, '');
}
```

Then unnest. There are other opportunities to flip branches and unnest below as well.

**Review by zanderso — APPROVED** [review]
lgtm w/ nit
